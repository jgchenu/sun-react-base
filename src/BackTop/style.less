@import '~@/styles/index';

@back-top-prefix-cls: ~'@{sun-prefix}-back-top';

.@{back-top-prefix-cls} {
  position: fixed;
  z-index: @back-top-z-index;
  right: 100px;
  bottom: 50px;
  width: 40px;
  height: 40px;
  cursor: pointer;

  &-inner {
    display: inline-block;

    &-icon {
      display: flex;
      width: @back-top-icon-width;
      height: @back-top-icon-height;
      align-items: center;
      justify-content: center;
      -webkit-appearance: none;
      background-color: @back-top-bg;
      border-radius: 50%;
      box-shadow: @default-shadow;
      color: @secondary;
      font-size: @back-top-icon-font-size;
      outline: none;

      i {
        display: block;
      }

      &:hover {
        border-color: @primary;
        background-color: @primary;
        color: @back-top-icon-hover-color;
      }
    }
  }
}
