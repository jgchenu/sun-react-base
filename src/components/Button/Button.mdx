---
name: Button
route: /button
---

import "../../styles/index.less";
import Button from "./Button.tsx";
import { Playground, Props } from "docz";

# Props Uasge

<Props of={Button} />

# Default Button

<Playground>
  <Button>default button</Button>
</Playground>

# Primary Button

<Playground>
  <Button btnType="primary">primary button</Button>
</Playground>
