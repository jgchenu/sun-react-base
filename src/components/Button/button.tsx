import React, { FC, ButtonHTMLAttributes, AnchorHTMLAttributes } from "react";
import classNames from "classnames";

export type ButtonSize = "lg" | "sm";
export type ButtonType = "primary" | "default" | "danger" | "link";

interface BaseButtonProps {
  className?: string;
  /** 禁用 */
  disabled?: boolean;
  /** 尺寸 */
  size?: ButtonSize;
  /** 类型 */
  btnType?: ButtonType;
  /** link类型button的href */
  href?: string;
}
type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLElement>;
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>;

export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>;

const Button: FC<ButtonProps> = (props) => {
  const {
    btnType,
    className,
    disabled,
    size,
    children,
    href,
    ...restProps
  } = props;
  const classes = classNames("btn", className, {
    [`btn-${btnType}`]: btnType,
    [`btn-${size}`]: size,
    disabled: btnType === "link" && disabled,
  });
  if (btnType === "link") {
    return (
      <a className={classes} href={href} {...restProps}>
        {children}
      </a>
    );
  } else {
    return (
      <button className={classes} disabled={disabled} {...restProps}>
        {children}
      </button>
    );
  }
};

Button.defaultProps = {
  disabled: false,
  btnType: "default",
};

export default Button;
