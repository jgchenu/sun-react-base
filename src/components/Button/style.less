.btn {
  position: relative;
  display: inline-block;
  border: @btn-border-width solid transparent;
  background-image: none;
  box-shadow: @btn-box-shadow;
  color: @body-color;
  cursor: pointer;
  font-weight: @btn-font-weight;
  line-height: @btn-line-height;
  text-align: center;
  transition: @btn-transition;
  vertical-align: middle;
  white-space: nowrap;
  .button-size(
    @btn-padding-y,
    @btn-padding-x,
    @btn-font-size,
    @border-radius
  );

  &.disabled,
  &[disabled] {
    box-shadow: none;
    cursor: not-allowed;
    opacity: @btn-disabled-opacity;

    > * {
      pointer-events: none;
    }
  }
}

.btn-lg {
  .button-size(
    @btn-padding-y-lg,
    @btn-padding-x-lg,
    @btn-font-size-lg,
    @btn-border-radius-lg
  );
}

.btn-sm {
  .button-size(
    @btn-padding-y-sm,
    @btn-padding-x-sm,
    @btn-font-size-sm,
    @btn-border-radius-sm
  );
}

.btn-primary {
  .button-style(@primary, @primary, @white);
}

.btn-danger {
  .button-style(@danger, @danger, @white);
}

.btn-default {
  .button-style(
    @white,
    @gray-400,
    @body-color,
    @white,
    @primary,
    @primary
  );
}

.btn-link {
  box-shadow: none;
  color: @btn-link-color;
  font-weight: @font-weight-normal;
  text-decoration: @link-decoration;

  &:hover {
    color: @btn-link-hover-color;
    text-decoration: @link-hover-decoration;
  }

  &:focus,
  &.focus {
    box-shadow: none;
    text-decoration: @link-hover-decoration;
  }

  &:disabled,
  &.disabled {
    color: @btn-link-disabled-color;
    pointer-events: none;
  }
}
