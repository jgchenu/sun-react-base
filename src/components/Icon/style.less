.icon-primary {
  color: @primary;
}

.icon-secondary {
  color: @secondary;
}

.icon-success {
  color: @success;
}

.icon-info {
  color: @info;
}

.icon-warning {
  color: @warning;
}

.icon-danger {
  color: @danger;
}

.icon-light {
  color: @light;
}

.icon-dark {
  color: @dark;
}
