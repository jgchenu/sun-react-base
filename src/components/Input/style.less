.sun-input-wrap {
  position: relative;
  display: flex;
  width: 100%;
  margin-bottom: 30px;

  .icon-wrap {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    width: 35px;
    height: 100%;
    align-items: center;
    justify-content: center;
    color: @input-color;

    svg {
      color: @input-placeholder-color;
    }
  }
}

.sun-input-inner {
  width: 100%;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-border-color;
  background-clip: padding-box;
  background-color: @input-bg;

  // Note: This has no effect on <select>s in some browsers, due to the limited stylability of `<select>`s in CSS.
  border-radius: @input-border-radius;
  box-shadow: @input-box-shadow;
  color: @input-color;
  font-family: @input-font-family;
  font-size: @input-font-size;
  font-weight: @input-font-weight;
  line-height: @input-line-height;
  transition: @input-transition;

  &:focus {
    border-color: @input-focus-border-color;
    background-color: @input-focus-bg;
    box-shadow: @input-focus-box-shadow;
    color: @input-focus-color;
    outline: 0;
    // Avoid using mixin so we can pass custom focus shadow properly
  }

  &::placeholder {
    color: @input-placeholder-color;
    // Override Firefox's unusual default opacity; see https://github.com/twbs/bootstrap/pull/11526.
    opacity: 1;
  }

  &:disabled,
  &[readonly] {
    border-color: @input-disabled-border-color;
    background-color: @input-disabled-bg;
    // iOS fix for unreadable disabled content; see https://github.com/twbs/bootstrap/issues/11655.
    opacity: 1;
  }
}

.icon-wrap + .sun-input-inner {
  padding-right: 35px;
}

.sun-input-group-prepend,
.sun-input-group-append {
  display: flex;
  align-items: center;
  padding: @input-padding-y @input-padding-x;
  border: @input-border-width solid @input-group-addon-border-color;
  margin-bottom: 0; // Allow use of <label> elements by overriding our default margin-bottom
  background-color: @input-group-addon-bg;
  border-radius: @input-border-radius;
  color: @input-group-addon-color;
  font-size: @input-font-size; // Match inputs
  font-weight: @font-weight-normal;
  line-height: @input-line-height;
  text-align: center;
  white-space: nowrap;
}

.input-size-sm .sun-input-inner {
  padding: @input-padding-y-sm @input-padding-x-sm;
  border-radius: @input-border-radius-sm;
  font-size: @input-font-size-sm;
}

.input-size-lg .sun-input-inner {
  padding: @input-padding-y-lg @input-padding-x-lg;
  border-radius: @input-border-radius-lg;
  font-size: @input-font-size-lg;
}

.sun-input-group-append + .btn {
  padding: 0;
  border: 0;
}

.input-group > .sun-input-group-prepend,
.input-group.input-group-append > .sun-input-inner {
  .border-right-radius(0);
}

.input-group > .sun-input-group-append,
.input-group.input-group-prepend > .sun-input-inner {
  .border-left-radius(0);
}
