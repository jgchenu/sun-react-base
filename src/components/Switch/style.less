.sun-switch {
  position: relative;
  display: inline-block;
  min-width: @switch-width;
  height: @switch-height;
  box-sizing: border-box;
  background-color: @switch-bg-color;
  border-radius: @switch-width;
  color: @gray-700;
  cursor: pointer;
  font-size: @font-size-base;
  line-height: @switch-height;
  user-select: none;
  vertical-align: middle;

  &::after {
    position: absolute;
    top: (@switch-height - @switch-circle-width) / 2;
    left: @switch-offset;
    width: @switch-circle-width;
    height: @switch-circle-width;
    background-color: #fff;
    border-radius: 50%;
    box-shadow: @default-shadow;
    content: " ";
    transition: @switch-animation;
  }

  .sun-switch-inner {
    margin-left: 24px;
    color: #fff;
    font-size: @font-size-base - 2px;
  }

  &.sun-switch-checked {
    background-color: @switch-color;

    &::after {
      left: 100%;
      margin-left: -@switch-offset;
      transform: translate3d(-100%, 0, 0);
    }

    .sun-switch-inner {
      margin-left: 6px;
    }

    .sun-switch-loading {
      &-icon {
        left: 100%;
        margin-left: -@switch-circle-width;
        color: @switch-color;
        opacity: 0.8;
        transform: translate3d(-100%, 0, 0);
      }
    }
  }

  &.sun-switch-disabled,
  &.sun-switch-loading {
    background-color: fade(@switch-bg-color, 10%);
    cursor: not-allowed;

    &.sun-switch-checked {
      background-color: fade(@primary, 40%);
    }
  }
}
