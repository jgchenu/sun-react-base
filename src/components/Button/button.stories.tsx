import React from "react";
import Button, { ButtonProps } from "./button";
import { action } from "@storybook/addon-actions";

// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from "@storybook/react/types-6-0";

export default {
  title: "Button",
  component: Button,
} as Meta;

const Template: Story<ButtonProps> = (args) => (
  <Button {...args} onClick={action("click")} />
);

export const Default = Template.bind({});
Default.args = {
  children: "default button",
};

export const Primary = Template.bind({});
Primary.args = {
  btnType: "primary",
  children: "default button",
};

export const Danger = Template.bind({});
Danger.args = {
  btnType: "danger",
  children: "danger button",
};

export const Large = Template.bind({});
Large.args = {
  size: "lg",
  children: "Large Button",
};

export const Small = Template.bind({});
Small.args = {
  size: "sm",
  children: "Small Button",
};
