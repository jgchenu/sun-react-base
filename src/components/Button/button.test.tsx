import React from "react";
import renderer from "react-test-renderer";
import Button, { ButtonProps } from "./Button";
import { mount } from "enzyme";

import "./style.less";

const defaultProps = {
  onClick: jest.fn(),
};

const testProps: ButtonProps = {
  btnType: "primary",
  size: "lg",
  className: "klass",
};

const disabledProps: ButtonProps = {
  disabled: true,
  onClick: jest.fn(),
};

describe("test Button component", () => {
  it("Button should render correct html", () => {
    const tree = renderer
      .create(<Button {...defaultProps}>Nice</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it("should render the correct default button", () => {
  //   const wrapper = mount(<Button {...defaultProps}>Nice</Button>);
  //   const element = wrapper.getByText("Nice") as HTMLButtonElement;
  //   expect(element).toBeInTheDocument();
  //   expect(element.tagName).toEqual("BUTTON");
  //   expect(element).toHaveClass("btn btn-default");
  //   expect(element.disabled).toBeFalsy();
  //   fireEvent.click(element);
  //   expect(defaultProps.onClick).toHaveBeenCalled();
  // });
  // it("should render the correct component based on different props", () => {
  //   const wrapper = render(<Button {...testProps}>Nice</Button>);
  //   const element = wrapper.getByText("Nice");
  //   expect(element).toBeInTheDocument();
  //   expect(element).toHaveClass("btn-primary btn-lg klass");
  // });
  // it("should render a link when btnType equals link and href is provided", () => {
  //   const wrapper = render(
  //     <Button btnType="link" href="http://dummyurl">
  //       Link
  //     </Button>
  //   );
  //   const element = wrapper.getByText("Link");
  //   expect(element).toBeInTheDocument();
  //   expect(element.tagName).toEqual("A");
  //   expect(element).toHaveClass("btn btn-link");
  // });
  // it("should render disabled button when disabled set to true", () => {
  //   const wrapper = render(<Button {...disabledProps}>Nice</Button>);
  //   const element = wrapper.getByText("Nice") as HTMLButtonElement;
  //   expect(element).toBeInTheDocument();
  //   expect(element.disabled).toBeTruthy();
  //   fireEvent.click(element);
  //   expect(disabledProps.onClick).not.toHaveBeenCalled();
  // });
});
