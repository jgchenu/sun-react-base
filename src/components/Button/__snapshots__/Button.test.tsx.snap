// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Button component Button should render correct html 1`] = `
<button
  className="btn btn-default"
  disabled={false}
  onClick={[MockFunction]}
>
  Nice
</button>
`;
